nexus/src/app/background/instance_watcher.rs

```diff
@@ -0,0 +1,593 @@
+                    Ok(batch) => batch,
+                    Err(e) => {
+                        slog::error!(
+                            opctx.log,
+                            "sled instances by sled agent query failed: {e}"
+                        );
+                        return serde_json::json!({ "error": e.to_string() });
+                    }
+                };
+                paginator = p.found_batch(&batch, &|(sled, _, _, _)| sled.id());
+
+                // When we iterate over the batch of sled instances, we pop the
+                // first sled from the batch before looping over the rest, to
+                // insure that the initial sled-agent client is created first,
+                // as we need the address of the first sled to construct it.
+                // We could, alternatively, make the sled-agent client an
+                // `Option`, but then every subsequent iteration would have to
+                // handle the case where it's `None`, and I thought this was a
+                // bit neater...
+                let mut batch = batch.into_iter();
+                if let Some((mut curr_sled, instance, vmm, project)) = batch.next() {
+                    let mut client = mk_client(&curr_sled);
+                    let target = VirtualMachine {
+                        instance_id: instance.id(),
+                        silo_id: project.silo_id,
+                        project_id: project.id(),
+                        vmm_id: vmm.id,
+                        sled_agent_id: curr_sled.id(),
+                        sled_agent_ip: (*curr_sled.address().ip()).into(),
+                        sled_agent_port: curr_sled.address().port(),
+                    };
+                    tasks.spawn(self.check_instance(opctx, &client, target));
+
+                    for (sled, instance, vmm, project) in batch {
+                        // We're now talking to a new sled agent; update the client.
+                        if sled.id() != curr_sled.id() {
+                            client = mk_client(&sled);
+                            curr_sled = sled;
+                        }
+                        let target = VirtualMachine {
+                            instance_id: instance.id(),
+                            silo_id: project.silo_id,
+                            project_id: project.id(),
+                            vmm_id: vmm.id,
+                            sled_agent_id: curr_sled.id(),
+                            sled_agent_ip: (*curr_sled.address().ip()).into(),
+                            sled_agent_port: curr_sled.address().port(),
+                        };
+                        tasks.spawn(self.check_instance(opctx, &client, target));
+                    }
+                }
+            }
+
+            // All requests fired off! While we wait for them to come back,
+            // let's prune old instances.
+            let pruned = self.metrics.lock().unwrap().prune();
```

Comment:
IIUC, here we prune instances based on whether they've been marked touched by the tasks we spawn above. Does that mean we're racing them? For example, suppose we have the following:

- We spawn the checking task, it marks the instance touched.
- The prune call swaps that with `false` so it's marked "untouched"
- On the next pass, we spawn the checking task, but before it runs...
- We get here, pruning that still-untouched instance.

I think the actual data we report to ClickHouse will be handled correctly, in either case, but it does mean there will appear to be many restarts of that cumulative timeseries. Logically, it seems like we should be using the database to tell us which instances to report data for: if we have it in our list from the database, keep it, otherwise prune it. That might not play well with the pagination you have here, but I think it could be made to work and avoids the race above.

Reply:
After thinking about this for a minute, I don't actually believe this is racy.
The checking task itself does *not* actually mutate the metrics, it just returns
a value to *this* task, which updates the metric and sets the touched flag. This is the only task that
mutates the metrics, and there can only be one of these tasks executing at any
given time, as activating the background task borrows it mutably. And, we won't
activate the background task again until *all* in-flight checks have completed,
because this task waits for all tasks spawned on the `JoinSet` to complete
before this future completes. There would be
a potential race if the checking tasks were responsible for actually updating
the metrics, but they're not.

I think there *is* an issue here, though, but it's much simpler: the issue is
just that we should call `prune` *after* all tasks spawned on the `JoinSet` have
completed, rather than before.

Reply:
> the checking task itself does not actually mutate the metrics

Ah, I didn't grok that, thanks.

> we should call prune after all tasks spawned on the JoinSet have completed

I think that makes sense. It seems logically equivalent to pruning any instances that are not returned from the database, right? If so, then that seems good to me.